### ISSUE TITLE
met a problem when handle websocket connection

### ISSUE THREAD

**Issue by hanzhe179** [pomerium/pomerium#3456]
Hi Experts, 

Hope if anyone has every met same issue can help me on this if any configuration mistakes.
I am trying to use Pomerium to protect a website which serve websocket connection. But after the main login authentication past and the websocket service is brought up for the session, then failed on ws connection.
The error messages like below:

Jun 29 11:33:28 eoc-gate pomerium[109648]: {"level":"info","service":"envoy","upstream-cluster":"route-cf1d0de8a18fcc26","method":"GET","authority":"login.example.com","path":"/websocket/hanzhe/10/YzA3ZjQ0MmQ2MDFhMWNhOTMwNDNjZmQ4OWRhMTVkNjk2NzE1MThmOWRkMWMwOGNkYWZhNTk5YzYzN2RmYTI0MjExMTM2ZWQ4NDQ5OGRmMjM3ZjkzZGFkYWIwMDEwZDBmZTFhMjliMWI5ZGI3ZjE3YjA3NTIwZDFmNzNlYTdmMjkkMTY1NjQ3MzYwNyQ0MTUwYjRjMTk4NjM1ZDU3YWEyZTg4MGFkNzU4YzlkMTQ5ZDJiMGRiN2M2MTEzYmZiZDI0OTVjN2RjNWUzMmNlZGM5YTdiZDRkZjc3ODBjMmM0MzM3NWFkMzlmZTkwZDVhNjc3ODRiZTVjMTFkN2E1MDBjNmVkMmVmNTAwMzkzMQ==","user-agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.3 Safari/605.1.15","referer":"","forwarded-for":"180.115.155.36","request-id":"b2e1ed22-13c5-4f90-9cab-4ecea6fea660","duration":23.731938,"size":95,"response-code":503,"response-code-details":"upstream_reset_before_response_started{connection_termination}","time":"2022-06-29T11:33:28+08:00","message":"http-request"}

On the apps server which serve this, the log message say that it can't extract the client IP address.

However, when using Nginix, the whole connection flow works when set proxy_set_header for variable Upgrade and Connection.

Thanks!

**Comment by hanzhe179** [comment]
Hi @calebdoxsey, thank you very much for the answers on this. 

I tried to add the option xff_num_trusted_hops as below. Also tried option skip_xff_append as true. But both no luck.

xff_num_trusted_hops: 2
routes:
  - from: https://login.example.com
    to: https://10.0.0.127
    allow_websockets: true
    allow_any_authenticated_user: true
    tls_server_name: login.example.com
    preserve_host_header: true

I saw there is another issue mentioned that meeting issue when using websocket with upstream as https.
https://github.com/pomerium/pomerium/issues/2388
Should I set two routes to match the wss service? 

Also, the error of my app which running on upstream as below. The OS is CentOS 7.9. 
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1] ----------------------------------------
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1] Exception happened during processing of request from ('::ffff:10.0.0.1', 41480, 0, 0)
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1] Traceback (most recent call last):
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]   File "/usr/lib64/python2.7/SocketServer.py", line 568, in process_request
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]     self.finish_request(request, client_address)
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]   File "/opt/thinlinc/sbin/tlwebaccess", line 613, in finish_request
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]     ForkingTLSServer . finish_request ( self , request , client_address )
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]   File "/opt/thinlinc/modules/thinlinc/tlstunnel.py", line 81, in finish_request
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]     self . TLSRequestHandlerClass ( request , client_address , self )
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]   File "/opt/thinlinc/modules/thinlinc/websocket.py", line 102, in __init__
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]     SimpleHTTPRequestHandler.__init__(self, req, addr, server)
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]   File "/usr/lib64/python2.7/SocketServer.py", line 649, in __init__
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]     self.handle()
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]   File "/opt/thinlinc/sbin/tlwebaccess", line 83, in handle
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]     thinlinc . websocket . WebSocketRequestHandler . handle ( self )
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]   File "/opt/thinlinc/modules/thinlinc/websocket.py", line 543, in handle
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]     SimpleHTTPRequestHandler.handle(self)
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]   File "/usr/lib64/python2.7/BaseHTTPServer.py", line 340, in handle
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]     self.handle_one_request()
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]   File "/usr/lib64/python2.7/BaseHTTPServer.py", line 328, in handle_one_request
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]     method()
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]   File "/opt/thinlinc/sbin/tlwebaccess", line 209, in do_GET
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]     if not self . handle_websocket ( ) :
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]   File "/opt/thinlinc/sbin/tlwebaccess", line 244, in handle_websocket
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]     if not self . check_websocket_authorization ( OOO00 [ 3 ] ) :
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]   File "/opt/thinlinc/sbin/tlwebaccess", line 412, in check_websocket_authorization
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1]     OOoO , OO0O000 = oooOo0OOOoo0 . split ( "=" , 1 )
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1] ValueError: need more than 1 value to unpack
2022-07-10 20:38:14 ERROR tlwebaccess[2901]: [::ffff:10.0.0.1] ----------------------------------------

**Comment by hanzhe179** [comment]
finally found the problem, the upstream service somehow got the Cookie from header as below:

auth_cookie=8741b80b3a3488ea15f846dd2599a757de77c9b31420f5f2d88e13d06d975678c3ae25cfe2a57cab5740619a8670a6e215277394bd2c34ce9220cb9e756256b7$1662724331$3d742c765c9bfccf15aa2aaede9dfbbfb0e583df34a1527d26fc56df750fd8cd2faa36bdfd1bbfc00f9d675d17b21e9a6db7267730a2ce0c3b6ea94429bb7227;

There is a ";" at the end, when program handle this, it splits whole thing by ";", so got second entry as "". Then it failed with wrong syntax. I changed the app program by removing that ";", and it works.

This doesn't happen when using simple Nginx as proxy.

Anyway, thank you all! Closing this issue.
